metrics/src/metrics.rs

```diff
@@ -0,0 +1,272 @@
+};
+
+pub use observability_deps::opentelemetry::KeyValue;
+
+const RED_REQUEST_STATUS_LABEL: &str = "status";
+
+/// Possible types of RED metric observation status.
+///
+/// Ok          - an observed request was successful.
+/// ClientError - an observed request was unsuccessful but it was not the fault
+///               of the observed service.
+/// Error       - an observed request failed and it was the fault of the
+///               service.
+///
+/// What is the difference between `ClientError` and `Error`? The difference is
+/// to do where the failure occurred. When thinking about measuring SLOs like
+/// availability it's necessary to calculate things like:
+///
+///    Availability = 1 - (failed_requests / all_valid_requests)
+///
+/// `all_valid_requests` includes successful requests and any requests that
+/// failed but not due to the fault of the service (e.g., client errors).
+///
+/// It is useful to track the components of `all_valid_requests` separately so
+/// operators can also monitor external errors (client_error) errors to help
+/// improve their APIs or other systems.
+#[derive(Debug)]
+pub enum RedRequestStatus {
+    Ok,
+    ClientError,
+    Error,
+}
+
+impl Display for RedRequestStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Ok => write!(f, "ok"),
+            Self::ClientError => write!(f, "client_error"),
+            Self::Error => write!(f, "error"),
+        }
+    }
+}
+
+#[derive(Debug)]
+/// A REDMetric is a metric that tracks requests to some resource.
+///
+/// The [RED methodology](https://www.weave.works/blog/the-red-method-key-metrics-for-microservices-architecture/)
+/// stipulates you should track three key measures:
+///
+///  - Request Rate: (total number of requests / second);
+///  - Error Rate: (total number of failed requests / second);
+///  - Duration: (latency distributions for the various requests)
+///
+/// Using a `REDMetric` makes following this methodology easy because it handles
+/// updating the three components for you.
+pub struct RedMetric {
+    default_labels: Vec<KeyValue>,
+    requests: OTCounter<u64>,
+    duration: OTHistorgram<f64>,
+}
+
+impl RedMetric {
+    pub(crate) fn new(
+        requests: OTCounter<u64>,
+        duration: OTHistorgram<f64>,
+        mut default_labels: Vec<KeyValue>,
+    ) -> Self {
+        // TODO(edd): decide what to do if `labels` contains
+        // RED_REQUEST_STATUS_LABEL.
+        // RedMetric always has a status label.
+        default_labels.insert(
+            0,
+            KeyValue::new(RED_REQUEST_STATUS_LABEL, RedRequestStatus::Ok.to_string()),
+        );
+
+        Self {
+            requests,
+            duration,
+            default_labels,
+        }
+    }
+
+    /// Returns a new observation that will handle timing and recording an
+    /// observation the metric is tracking.
+    pub fn observation(
+        &'_ self,
```

Comment:
I don't think this `'_` on `self` is needed, it compiles for me without it...

```suggestion
        &self,
```